moler/connection_observer.py

```diff
@@ -210,15 +211,46 @@ def get_unraised_exceptions(remove=True):
                 return list_of_exceptions
 
     @staticmethod
-    def _append_to_not_raised_exceptions(exception):
+    def _change_unraised_exception(new_exception, observer):
         with ConnectionObserver._exceptions_lock:
-            ConnectionObserver._not_raised_exceptions.append(exception)
+            old_exception = observer._exception
+            ConnectionObserver._log_unraised_exceptions(observer)
+            if old_exception:
+                observer._log(logging.DEBUG,
+                              "'{}.{}' has overwritten exception. From '{}.{}' ({}) to '{}.{}' ({}).".format(
+                                  observer.__class__.__module__,
+                                  observer.__class__.__name__,
+                                  old_exception.__class__.__module__,
+                                  old_exception.__class__.__name__,
+                                  old_exception,
+                                  new_exception.__class__.__module__,
+                                  new_exception.__class__.__name__,
+                                  new_exception,
+                              ))
+                if old_exception in ConnectionObserver._not_raised_exceptions:
+                    ConnectionObserver._not_raised_exceptions.remove(old_exception)
+                else:
+                    observer._log(logging.DEBUG,
+                                  "'{}.{}': cannot find exception '{}.{}' '{}' in _not_raised_exceptions.".format(
+                                      observer.__class__.__module__,
+                                      observer.__class__.__name__,
+                                      old_exception.__class__.__module__,
+                                      old_exception.__class__.__name__,
+                                      old_exception,
+                                  ))
+                    ConnectionObserver._log_unraised_exceptions(observer)
+
+            ConnectionObserver._not_raised_exceptions.append(new_exception)
+            observer._exception = new_exception
 
     @staticmethod
-    def _remove_from_not_raised_exceptions(exception):
-        with ConnectionObserver._exceptions_lock:
-            if exception in ConnectionObserver._not_raised_exceptions:
-                ConnectionObserver._not_raised_exceptions.remove(exception)
+    def _log_unraised_exceptions(observer):
+        observer._log(logging.DEBUG, "list length: {}".format(len(ConnectionObserver._not_raised_exceptions)))
+        observer._log(logging.DEBUG, "list: {}".format(ConnectionObserver._not_raised_exceptions))
+        i = 0
+        for item in ConnectionObserver._not_raised_exceptions:
+            observer._log(logging.DEBUG, "{}: {}".format(i, item))
+            i += 1
```

Comment:
Try to use enumerate() like here:
https://stackoverflow.com/questions/522563/accessing-the-index-in-for-loops

Reply:
done